include/oneapi/dpl/pstl/unseq_backend_simd.h

```diff
@@ -879,6 +879,41 @@ __simd_remove_if(_RandomAccessIterator __first, _DifferenceType __n, _UnaryPredi
     }
     return __current + __cnt;
 }
+
+template <typename _Iterator1, typename _Iterator2, typename _Iterator3, typename _Comp>
+std::pair<_Iterator1, _Iterator2>
+__simd_merge(_Iterator1 __x, _Iterator1 __x_e, _Iterator2 __y, _Iterator2 __y_e, _Iterator3 __i, _Iterator3 __j,
```

Comment:
Does this actually vectorize?  I can't imagine the compiler is able to do anything at all here, all operations are conditionals and dependent on the previous iteration(s).
Unless you actually see vectorization here (which I'd be very interested to see), I'd recommend just having a single implementation of merge, and ignoring SIMD tag for this pattern.

Reply:
> Does this actually vectorize?

Looking at the perf result - yes.

Reply:
Looking at the different options in [godbolt](https://godbolt.org/z/rv3eT98nK), I don't think it is vectorizing it, but rather it is a faster implementation that is able to do some unrolling.

I'd recommend to do at least one of the following:  
1) removing the SIMD pragma and recollecting timing results for SIMD (my guess is no change)
2) rewrite the non-vector function with a similar structure and recollect timing results (my guess is a speedup for non-vector version)
3) Inspect the godbolt link above for ASM

With these, if you can confirm my suspicions, then I'd prefer to use a single performant implementation which is structured like the SIMD impl, but which does not attempt to vectorize.

Reply:
You can check the vectorization report to see if the pragma really works or is ignored, and decide based on that.

Reply:
> You can check the vectorization report to see if the pragma really works or is ignored, and decide based on that.

icx 2025:
```
<source>:76:5: warning: loop not vectorized: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering [-Wpass-failed=transform-warning]
```

Reply:
OK, so I'd suggest using something like this impl for both vector and non vector implementations (without the pragma).

Reply:
Actually, I leave just one serial implementation - `__serial_merge_out_lim`

Reply:
sounds like a plan.